src/main/java/problems/impl/TransactionalStack.java

```diff
@@ -0,0 +1,115 @@
+package problems.impl;
+
+import problems.exceptions.TransactionRolledbackException;
+
+import java.util.EmptyStackException;
+import java.util.Stack;
+
+public class TransactionalStack<T> {
+    private final Stack<T> stack = new Stack<>();
+    private final Stack<TransactionEvent<T>> transactionEvents = new Stack<>();
+    private Stack<TransactionEvent<T>> rolledbackTransactionEvents = new Stack<>();
+
+
+    private static class TransactionEvent<E> {
+        private enum Type {
+            BEGIN,
+            PUSH,
+            POP
+        }
+
+        private final Type type;
+        private final E value;
+
+        public TransactionEvent(Type type, E value) {
+            this.type = type;
+            this.value = value;
+        }
+    }
+
+    public void push(T value) {
+        stack.push(value);
+
+        if (!transactionEvents.isEmpty() && rolledbackTransactionEvents.empty()) {
+            transactionEvents.add(new TransactionEvent<>(TransactionEvent.Type.PUSH, value));
+        }
+    }
+
+    public T pop() {
+        T value;
+
+        try {
+            value = stack.pop();
+        } catch (EmptyStackException e) {
+            if (transactionEvents.empty()) {
+                throw e;
+            }
+
+            rollback();
+
+            throw new TransactionRolledbackException("Cannot pop an empty stack. Rolled back changes.");
+        }
+
+        if (!transactionEvents.isEmpty() && rolledbackTransactionEvents.empty()) {
+            transactionEvents.add(new TransactionEvent<>(TransactionEvent.Type.POP, value));
+        }
+
+        return value;
+    }
+
+    public int getSize() {
+        return stack.size();
+    }
+
+    public T peek() {
+        try {
+            return stack.peek();
+        } catch (EmptyStackException e) {
+            if (transactionEvents.empty()) {
+                throw e;
+            }
+
+            rollback();
+
+            throw new TransactionRolledbackException("Cannot peek an empty stack. Rolled back changes.");
+        }
+    }
+
+    public void begin() {
+        transactionEvents.add(new TransactionEvent<>(TransactionEvent.Type.BEGIN, null));
+    }
+
+    public void commit() {
+        while (!transactionEvents.empty()) {
+            TransactionEvent<T> transactionEvent = transactionEvents.pop();
+            if (transactionEvent.type == TransactionEvent.Type.BEGIN) {
+                return;
+            }
+        }
+    }
+
+    public void rollback() {
+        while (!transactionEvents.empty()) {
+            TransactionEvent<T> transactionEvent = transactionEvents.pop();
+            rolledbackTransactionEvents.push(transactionEvent);
+
+            switch (transactionEvent.type) {
+                case POP: {
+                    push(transactionEvent.value);
+                    break;
+                }
+
+                case PUSH: {
+                    pop();
+                    break;
+                }
+
+                default: {
```

Comment:
I would throw a `RuntimeException` to protect agains future changes. Since the values are coming from an enum, the only possible way of it happening is when you (or someone) adds a new value to the enum and don't handle it appropriately.